src/keylessBackup/encryption.ts

```diff
@@ -0,0 +1,107 @@
+      salt: 'some fixed salt',
+      info: 'valora.keylessBackup.deriveKeyFromKeyShares',
+      hash: 'SHA-256',
+    }),
+  ])
+
+  // Using futoin-hkdf to derive a 256-bit key
+  return hkdf(combinedKeyShares, outputKeyBytes, {
+    salt: 'some fixed salt',
+    info: 'valora.keylessBackup.deriveKeyFromKeyShares',
+    hash: 'SHA-256',
+  })
+}
+
+/**
+ * Derives a 256-bit key using the HKDF method from two key shares then derives a secp256k1 pair from that.
+ *
+ * @param {string} keyshare1 - The first keyshare in utf8 format.
+ * @param {string} keyshare2 - The second keyshare in utf8 format.
+ */
+export function getSecp256K1KeyPair(
+  keyshare1: Buffer,
+  keyshare2: Buffer
+): { privateKey: Uint8Array; publicKey: Uint8Array } {
+  const derivedKey = deriveKeyFromKeyShares(keyshare1, keyshare2, 48) // 40 is the minimum for hashToPrivateKey
+  const privateKey = secp.utils.hashToPrivateKey(derivedKey)
+  const publicKey = secp.getPublicKey(privateKey, true)
+  return { privateKey, publicKey }
+}
+
+/**
+ * Encrypts a passphrase using AES-256-GCM encryption with a key derived from two key shares.
+ *
+ * @param {string} keyshare1 - The first keyshare used to derive the encryption key.
+ * @param {string} keyshare2 - The second keyshare used to derive the encryption key.
+ * @param {string} passphrase - The passphrase to encrypt.
+ * @returns {string} The encrypted passphrase, formatted as `nonce:encrypted:authTag`, all parts base64 encoded.
+ */
+export function encryptPassphrase(
+  keyshare1: Buffer,
+  keyshare2: Buffer,
+  passphrase: string
+): string {
+  const derivedKey = deriveKeyFromKeyShares(keyshare1, keyshare2)
+  const nonce = crypto.randomBytes(12) // GCM recommends a 12-byte nonce
+  const cipher = crypto.createCipheriv('aes-256-gcm', derivedKey, nonce)
+
+  let encrypted = cipher.update(passphrase, 'utf8', 'base64')
+  encrypted += cipher.final('base64')
+  const authTag = cipher.getAuthTag()
+
+  return `${nonce.toString('base64')}:${encrypted}:${authTag.toString('base64')}`
+}
+
+/**
+ * Decrypts an encrypted passphrase using AES-256-GCM decryption with a key derived from two key shares.
+ *
+ * @
+ * @param {string} keyshare1 - The first keyshare used to derive the decryption key.
+ * @param {string} keyshare2 - The second keyshare used to derive the decryption key.
+ * @param {string} encryptedData - The data to decrypt, formatted as `nonce:encrypted:authTag`, all parts base64 encoded.
+ * @returns {string} The decrypted passphrase.
+ */
+export function decryptPassphrase(
+  keyshare1: Buffer,
+  keyshare2: Buffer,
+  encryptedData: string
+): string {
+  const derivedKey = deriveKeyFromKeyShares(keyshare1, keyshare2)
+  const [nonceBase64, encrypted, authTagBase64] = encryptedData.split(':')
+  const nonce = Buffer.from(nonceBase64, 'base64')
+  const authTag = Buffer.from(authTagBase64, 'base64')
```

Comment:
What will happen in the (exceedingly unlikely) case that the authentication tag signals that the ciphertext has been tampered with? Does the crypto library throw?

Reply:
yep https://nodejs.org/api/crypto.html#deciphersetauthtagbuffer-encoding

> If no tag is provided, or if the cipher text has been tampered with, [decipher.final()](https://nodejs.org/api/crypto.html#decipherfinaloutputencoding) will throw, indicating that the cipher text should be discarded due to failed authentication